src/webgpu/api/validation/encoding/cmds/render/dynamic_state.spec.ts

```diff
@@ -130,71 +130,97 @@ class F extends ValidationTest {
+g.test('setViewport,exceeds_attachment_size')
+  .desc(`Test that the viewport can exceed the attachment size`)
+  .paramsSubcasesOnly([
+    { attachmentWidth: 3, attachmentHeight: 3 },
+    { attachmentWidth: 1024, attachmentHeight: 1024 },
+  ])
+  .fn(t => {
+    const { attachmentWidth, attachmentHeight } = t.params;
+    t.testViewportCall(
+      true,
+      { x: 0, y: 0, w: attachmentWidth + 1, h: attachmentHeight + 1, minDepth: 0, maxDepth: 1 },
+      { width: attachmentWidth, height: attachmentHeight, depthOrArrayLayers: 1 }
+    );
+  });
+
+g.test('setViewport,xy_rect_contained_in_bounds')
   .desc(
-    'Test that the rectangle defined by x, y, width, height must be contained in the attachments'
+    `Test that the rectangle defined by x, y, width, height must be contained in the maximum viewport bounds
+and that the viewport size cannot exceed the maximum.`
   )
   .paramsSubcasesOnly(u =>
-    u
-      .combineWithParams([
-        { attachmentWidth: 3, attachmentHeight: 5 },
-        { attachmentWidth: 5, attachmentHeight: 3 },
-        { attachmentWidth: 1024, attachmentHeight: 1 },
-        { attachmentWidth: 1, attachmentHeight: 1024 },
-      ])
-      .combineWithParams([
-        // Control case: a full viewport is valid.
-        { dx: 0, dy: 0, dw: 0, dh: 0 },
-
-        // Other valid cases with a partial viewport.
-        { dx: 1, dy: 0, dw: -1, dh: 0 },
-        { dx: 0, dy: 1, dw: 0, dh: -1 },
-        { dx: 0, dy: 0, dw: -1, dh: 0 },
-        { dx: 0, dy: 0, dw: 0, dh: -1 },
-
-        // Test with a small value that causes the viewport to go outside the attachment.
-        { dx: 1, dy: 0, dw: 0, dh: 0 },
-        { dx: 0, dy: 1, dw: 0, dh: 0 },
-        { dx: 0, dy: 0, dw: 1, dh: 0 },
-        { dx: 0, dy: 0, dw: 0, dh: 1 },
-      ])
+    u.combineWithParams([
+      // Control case: max viewport is valid.
+      { mx: 0, my: 0, dx: 0, dy: 0, dw: 0, dh: 0 },
+
+      // Other valid cases
+      { mx: -1, my: 0, dx: 0, dy: 0, dw: 0, dh: 0 },
+      { mx: -2, my: 0, dx: 0, dy: 0, dw: 0, dh: 0 },
+      { mx: 1, my: 0, dx: -1, dy: 0, dw: 0, dh: 0 },
+      { mx: 0, my: -1, dx: 0, dy: 0, dw: 0, dh: 0 },
+      { mx: 0, my: -2, dx: 0, dy: 0, dw: 0, dh: 0 },
+      { mx: 0, my: 1, dx: 0, dy: -1, dw: 0, dh: 0 },
+      { mx: 0, my: 0, dx: -1, dy: 0, dw: 0, dh: 0 },
+      { mx: 0, my: 0, dx: 1, dy: 0, dw: 0, dh: 0 },
+      { mx: 0, my: 0, dx: 0, dy: -1, dw: 0, dh: 0 },
+      { mx: 0, my: 0, dx: 0, dy: 1, dw: 0, dh: 0 },
+      { mx: 1, my: 0, dx: 0, dy: 0, dw: -1, dh: 0 },
+      { mx: 0, my: 1, dx: 0, dy: 0, dw: 0, dh: -1 },
+
+      // Cases that go outside the allowed bounds
+      { mx: -2, my: 0, dx: -1, dy: 0, dw: 0, dh: 0 },
+      { mx: 0, my: -2, dx: 0, dy: -1, dw: 0, dh: 0 },
+      { mx: 1, my: 0, dx: 0, dy: 0, dw: 0, dh: 0 },
+      { mx: 0, my: 1, dx: 0, dy: 0, dw: 0, dh: 0 },
+      { mx: 1, my: 0, dx: 1, dy: 0, dw: -1, dh: 0 },
+      { mx: 0, my: 1, dx: 0, dy: 1, dw: 0, dh: -1 },
```

Comment:
dw and dh seem to be unused?

I think if they were used then some of these cases would change their expected results.

Reply:
Done